sympy/core/equation.py

```diff
@@ -0,0 +1,342 @@
+    operator is recognized.
+
+    This class is intended to allow using the mathematical tools in SymPy to
+    rearrange equations and perform algebra in a stepwise fashion. In this
+    way more people can successfully perform algebraic rearrangements without
+    stumbling over missed details such as a negative sign.
+
+    Create an equation with the call ``Equation(lhs,rhs)``, where ``lhs`` and
+    ``rhs`` are any valid Sympy expression. ``Eqn(...)`` is a synonym for
+    ``Equation(...)``.
+
+    Examples
+    ========
+    >>> from sympy import var, Equation, Eqn, exp, log, integrate, Integral
+    >>> from sympy import solve
+    >>> a, b, c = var('a b c')
+    >>> str(Equation(a,b/c))
+    a=b/c
+    >>> t=Eqn(a,b/c)
+    >>> str(t)
+    a=b/c
+    >>> str(t*c)
+    a*c=b
+    >>> str(c*t)
+    a*c=b
+    >>> str(exp(t))
+    exp(a)=exp(b/c)
+    >>> str(exp(log(t)))
+    a=b/c
+
+    Integration can only be performed on one side at a time.
+    >>> q=Eqn(a*c,b/c)
+    >>> integrate(q,b,side='rhs')
+    b**2/(2*c)
+    >>> integrate(q,b,side='lhs')
+    a*b*c
+
+    Make a pretty statement of integration from an equation
+    >>> str(Eqn(Integral(q.lhs,b),integrate(q,b,side='rhs')))
+    Integral(a*c, b)=b**2/(2*c)
+
+    This is duplicated by the convenience function self.integ
+    >>> str(q.integ(b))
+    Integral(a*c, b)=b**2/(2*c)
+
+    SymPy's solvers do not understand these equations. They expect an
+    expression that the solver assumes = 0. Thus to use the solver the
+    equation must be rearranged so that all non-zero symbols are on one side.
+    Then just the non-zero symbolic side is passed to ``solve()``.
+    >>> t2 = t-t.rhs
+    >>> str(t2)
+    a-b/c=0
+    >>> solve(t2.lhs,c)
+    [b/a]
+    """
+
+    def __new__(cls, lhs, rhs, **kwargs):
+        check = kwargs.pop('check', True)
+        lhs = _sympify(lhs)
+        rhs = _sympify(rhs)
+        if check:
+            lsimp = lhs.simplify()
+            rsimp = rhs.simplify()
+            if lhs.is_number and rhs.is_number and lsimp != rsimp:
+                from warnings import warn
+                warnstr = '\nDid your really mean to define unequal numbers '
+                warnstr += str(lsimp) + ' and ' + str(rsimp) + ' as equal?\n'
+                warnstr += 'To suppress this warning include `check=False`'
+                warnstr += ' in the equation definition: '
+                warnstr += '`Eqn(lhs,rhs, check=False)`.'
+                warn(warnstr)
+        return super().__new__(cls, lhs, rhs)
+
+    @property
+    def lhs(self):
+        """
+        Returns the lhs of the equation.
+        """
+        return self.args[0]
+
+    @property
+    def rhs(self):
+        """
+        Returns the rhs of the equation.
+        """
+        return self.args[1]
+
+    def as_Boolean(self):
+        """
+        Converts the equation to an Equality.
+        """
+        from .relational import Equality
+        return Equality(self.lhs, self.rhs)
```

Comment:
May I suggest to rename this method to `as_relational`? This would be consistent with the same method of the class Set. Also, I would consider adding a parameter `cls_rel`, with default value to `Equality`, representing the desired output class for the relational. Suppose I start with an inequality, `ineq = x > 5`; I would like to add a term to both its sides. Then, I would convert it to Equation and perform the addition, `eq = Equation(*ineq.args) + something`; finally convert it back: `eq.as_relational(Gt)`. Obviously, it is left to the user to make sure the sign of the output inequality is consistent with the performed operation.

Reply:
The `as_relational` method on sets takes an argument so this wouldn't have the same signature. Adding a method with the same name and a different signature is confusing I think.

It's not hard to convert the `Equation` to an inequality if you really want to with `Lt(eq.lhs, eq.rhs)` and I think that makes it clearer to the user that it is their responsibility to ensure that it is mathematically correct to do that.

The reason for having `as_Boolean()` is that it opens the door to having an `as_Boolean()` method on other classes. For example if `Boolean` had a no-op `as_Boolean()` method then a function like `Piecewise` can simply call `as_Boolean()` on the relevant args without needing to do any other type checking.

Reply:
I see your points. On the other hand, as a user I intend an object of type Equality to be a relation, not a boolean; from a programmer point of view it's definitely both. In this respect, the name `as_Boolean` might be confusing.

Reply:
The distinction between Equation and Equality is the fact that Equality is a Boolean so if you found yourself wanting to convert between them then I think `as_Boolean` would make sense.

In any case mostly users wouldn't need this method because we can use `as_Boolean` in contexts where a Boolean is expected meaning that it happens implicitly from the user perspective.